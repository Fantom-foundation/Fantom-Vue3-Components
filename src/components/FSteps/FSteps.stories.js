import FSteps from './FSteps.vue';

export default {
    title: 'FSteps',
    component: FSteps,
};

export const Default = () => ({
    components: { FSteps },
    template: `
        <div>
            <FSteps :count="3" />
        </div>
    `,
});

export const Active = () => ({
    components: { FSteps },
    template: `
        <div>
            <FSteps :count="4" :active="2" />
            <FSteps :count="4" :active="20" />
            <FSteps :count="4" :active="0" />
        </div>
    `,
});

export const WithLabels = () => ({
    components: { FSteps },
    template: `
        <div>
            <FSteps :labels="labels" />
            <FSteps :labels="labels" :active="2" />
            <FSteps :labels="labels" :active="3" />
        </div>
    `,
    data() {
        return {
            labels: ['Step 1', 'Step 2', 'Step 3'],
        };
    },
});
