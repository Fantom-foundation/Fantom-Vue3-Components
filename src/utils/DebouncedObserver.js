import { debounce } from './index.js';

export class DebouncedObserver {
    /** Observer instance */
    #observer = null;
    /** Observer class */
    #Observer = null;
    /** Ignore first callback call */
    #firstCall = false;

    constructor(Observer = window.ResizeObserver) {
        this.#Observer = Observer;
    }

    /**
     * @param {HTMLElement} element
     * @param {function} callback
     * @param {number} debounceInterval In milliseconds. 0 means no debouncing.
     * @param {boolean} ignoreFirstCall
     * @param {Object} options
     */
    create({ element, callback, debounceInterval = 0, ignoreFirstCall = false, options = {} }) {
        this.destroy();

        if (this.#Observer) {
            this.#firstCall = true;

            const _callback = (entries) => {
                if (ignoreFirstCall) {
                    if (!this.#firstCall) {
                        callback(entries);
                    } else {
                        this.#firstCall = false;
                    }
                } else {
                    callback(entries);
                }
            };

            this.#observer = new this.#Observer(
                debounceInterval > 0 ? debounce(_callback, debounceInterval, true) : _callback
            );
            this.#observer.observe(element, options);
        }
    }

    destroy() {
        if (this.#observer !== null) {
            this.#observer.disconnect();
            this.#observer = null;
        }
    }
}
