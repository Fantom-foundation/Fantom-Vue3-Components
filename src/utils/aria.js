import { getAttr, setAttr } from './DOM.js';

const RECEIVE_FOCUS_FROM_ATTR = 'data-receive-focus-from';

const KEY_EVENTS = {
    keyup: null,
    keydown: null,
};

/**
 * @param {string} key Name of the key or 'Shift|Ctrl|Alt|Meta+key_name'
 * @param {KeyboardEvent} event
 * @return {boolean}
 */
export function isKey(key, event) {
    let isKey = false;
    const ks = key.split('+');
    let k = key;

    if (ks.length === 2) {
        k = ks[1];
    }

    if (event) {
        switch (event.key) {
            case 'Esc': // IE/Edge
            case 'Escape':
                isKey = k === 'Escape';
                break;
            case 'Up': // IE/Edge
            case 'ArrowUp':
                isKey = k === 'ArrowUp';
                break;
            case 'Right': // IE/Edge
            case 'ArrowRight':
                isKey = k === 'ArrowRight';
                break;
            case 'Left': // IE/Edge
            case 'ArrowLeft':
                isKey = k === 'ArrowLeft';
                break;
            case 'Down': // IE/Edge
            case 'ArrowDown':
                isKey = k === 'ArrowDown';
                break;
            case 'Spacebar': // IE/Edge
            case ' ':
                isKey = k === ' ';
                break;
            default:
                isKey = k === event.key;
        }

        if (isKey && ks.length === 2) {
            switch (ks[0]) {
                case 'Shift':
                    isKey = event.shiftKey;
                    break;
                case 'Ctrl':
                    isKey = event.ctrlKey;
                    break;
                case 'Alt':
                    isKey = event.altKey;
                    break;
                case 'Meta': // mac
                    isKey = event.metaKey;
                    break;
            }
        }
    }

    return isKey;
}

/**
 * Is element visible? Fast method, but not accurate, if fixed element is on page.
 * (window.getComputedStyle(el) (style.display === 'none') -> much slower).
 *
 * @param {HTMLElement} _elem
 * @return {boolean}
 */
export function isVisible(_elem) {
    return !!_elem && _elem.offsetParent !== null;
}

/**
 * @param {HTMLElement} _elem
 * @return {boolean}
 */
export function isFocusable(_elem) {
    let focusable = false;

    if (!_elem || _elem.disabled) {
        return false;
    }

    switch (_elem.nodeName) {
        case 'INPUT':
            focusable = _elem.type !== 'hidden' && _elem.type !== 'file';
            break;
        case 'A':
            focusable = !!_elem.href && _elem.rel !== 'ignore';
            break;
        case 'BUTTON':
        case 'SELECT':
        case 'TEXTAREA':
            focusable = true;
            break;
        default:
            focusable = false;
    }

    if (!focusable) {
        focusable = _elem.tabIndex > 0 || (_elem.tabIndex === 0 && _elem.getAttribute('tabIndex') !== null);
    }

    if (focusable) {
        focusable = isVisible(_elem);
    }

    return focusable;
}

/**
 * @param {HTMLElement} _elem
 * @return {HTMLElement|null}
 */
export function findFirstFocusableDescendant(_elem) {
    let focusableElem = null;
    let childNodes;
    let child;

    if (_elem) {
        childNodes = _elem.assignedNodes ? _elem.assignedNodes() : _elem.childNodes;

        // Edge/IE
        if (!childNodes && _elem.assignedNodes) {
            childNodes = _elem.childNodes;
        }

        if (childNodes) {
            for (let i = 0, len1 = childNodes.length; i < len1; i += 1) {
                child = childNodes[i];
                if (child.nodeType === 1) {
                    if (isFocusable(child)) {
                        focusableElem = child;
                        break;
                    } else {
                        focusableElem = findFirstFocusableDescendant(child);
                        if (focusableElem) {
                            break;
                        }
                    }
                }
            }
        }
    }

    return focusableElem;
}

/**
 * @param {HTMLElement} _elem
 * @return {HTMLElement|null}
 */
export function findLastFocusableDescendant(_elem) {
    let focusableElem = null;
    let childNodes;
    let child;

    if (_elem) {
        childNodes = _elem.assignedNodes ? _elem.assignedNodes() : _elem.childNodes;

        // Edge/IE
        if (!childNodes && _elem.assignedNodes) {
            childNodes = _elem.childNodes;
        }

        if (childNodes) {
            for (let i = childNodes.length - 1; i >= 0; i -= 1) {
                child = childNodes[i];
                if (child.nodeType === 1) {
                    // console.log('la', child.assignedNodes, (child.assignedNodes ? child.assignedNodes() : ''));
                    if (isFocusable(child)) {
                        // return child;
                        focusableElem = child;
                        break;
                    } else {
                        focusableElem = findLastFocusableDescendant(child);
                        if (focusableElem) {
                            break;
                        }
                    }
                }
            }
        }
    }

    return focusableElem;
}

/**
 * @param {string} _id
 */
export function returnFocus(_id) {
    const elem = document.querySelector(`[${RECEIVE_FOCUS_FROM_ATTR}="${_id}"]`);

    if (elem) {
        elem.removeAttribute(RECEIVE_FOCUS_FROM_ATTR);
        elem.focus();
    } else {
        document.body.focus();
    }
}

/**
 * Set `RECEIVE_FOCUS_FROM_ATTR` attribute to an element with focus.
 * @param {string} _id
 */
export function setReceiveFocusFromAttr(_id) {
    if (document.activeElement) {
        document.activeElement.setAttribute(RECEIVE_FOCUS_FROM_ATTR, _id);
    }
}

/**
 * Focus element by selector
 *
 * @param {HTMLElement} [wrapper]
 * @param {string} [selector]
 */
export function focusElem(wrapper = null, selector = '[data-focus]') {
    const eWrapper = wrapper || document.body;
    const elem = eWrapper.querySelector(selector);

    if (elem) {
        if (!isFocusable(elem)) {
            elem.setAttribute('tabindex', '-1');
        }

        elem.focus();
    }
}

/**
 *
 * @param {MouseEvent|KeyboardEvent} _event
 * @return {boolean}
 */
export function isAriaAction(_event) {
    if (!_event) {
        return false;
    }

    const eventType = _event.type;

    if (eventType === 'click') {
        return true;
    }

    return eventType in KEY_EVENTS && (isKey('Enter', _event) || isKey(' ', _event));
}

/**
 * Trap focus inside an element.
 *
 * @param {KeyboardEvent} _event
 * @param {HTMLElement} _elem
 * @param {{first: HTMLElement, last: HTMLElement}} _focusableElems
 */
export function focusTrap(_event, _elem, _focusableElems) {
    if (isKey('Tab', _event)) {
        if (!_focusableElems.first) {
            _focusableElems.first = findFirstFocusableDescendant(_elem);
        }
        if (!_focusableElems.last) {
            _focusableElems.last = findLastFocusableDescendant(_elem);
        }

        if (_event.target === _focusableElems.first && _event.shiftKey) {
            if (_focusableElems.last) {
                _event.preventDefault();
                _focusableElems.last.focus();
            }
        } else if (_event.target === _focusableElems.last && !_event.shiftKey) {
            if (_focusableElems.first) {
                _event.preventDefault();
                _focusableElems.first.focus();
            }
        }
    }
}

/**
 * @param {HTMLElement} _eTarget
 * @param {string} _selector
 * @return {HTMLElement|null}
 */
export function getFirstElement(_eTarget, _selector) {
    let elem = _eTarget && _eTarget.parentElement ? _eTarget.parentElement.firstElementChild : null;

    while (elem && !elem.matches(_selector)) {
        elem = elem.nextElementSibling;
    }

    return elem;
}

/**
 * @param {HTMLElement} _eTarget
 * @param {string} _selector
 * @return {HTMLElement|null}
 */
export function getLastElement(_eTarget, _selector) {
    let elem = _eTarget && _eTarget.parentElement ? _eTarget.parentElement.lastElementChild : null;

    while (elem && !elem.matches(_selector)) {
        elem = elem.previousElementSibling;
    }

    return elem;
}

/**
 *
 * @param {KeyboardEvent} event
 * @param {('horizontal' | 'vertical' | 'both')} [direction] Movement direction.
 * @param {boolean} [useHomeAndEnd] `Home` and `End` keys are enabled.
 * @return {string}
 */
function getMove({ event, direction = 'horizontal', useHomeAndEnd }) {
    let move = '';

    if (direction === 'horizontal' || direction === 'both') {
        if (isKey('ArrowRight', event)) {
            move = 'next';
        } else if (isKey('ArrowLeft', event)) {
            move = 'prev';
        }
    }

    if (!move && (direction === 'vertical' || direction === 'both')) {
        if (isKey('ArrowDown', event)) {
            move = 'next';
        } else if (isKey('ArrowUp', event)) {
            move = 'prev';
        }
    }

    if (!move) {
        if (useHomeAndEnd) {
            if (isKey('Home', event)) {
                move = 'first';
            } else if (isKey('End', event)) {
                move = 'last';
            }
        }

        if (isKey('PageUp', event)) {
            move = 'first';
        } else if (isKey('PageDown', event)) {
            move = 'last';
        }
    }

    return move;
}

/**
 *
 * @param {KeyboardEvent} _event
 * @param {string} _selector Item selector.
 * @param {('horizontal' | 'vertical' | 'both')} [_direction] Movement direction.
 * @param {boolean} [_circular] Circular keyboard navigation
 * @param {HTMLElement} [_target]
 * @param {boolean} [_focusElem] Focus found element.
 * @param {boolean} [_useHomeAndEnd] `Home` and `End` keys are enabled.
 * @return {HTMLElement|null} Next or previous element or null.
 */
export function keyboardNavigation({
    _event,
    _selector,
    _direction = 'horizontal',
    _circular = false,
    _target = null,
    _focusElem = true,
    _useHomeAndEnd = true,
}) {
    if (!_event || !(_event.type in KEY_EVENTS) || !_selector) {
        return null;
    }

    const eTarget = _target || _event.target.closest(_selector);
    let elem = null;
    let move = '';

    // console.log(_event.code);

    if (eTarget) {
        move = getMove({ event: _event, direction: _direction, useHomeAndEnd: _useHomeAndEnd });

        if (move === 'next') {
            elem = eTarget.nextElementSibling;
            while (elem && !elem.matches(_selector)) {
                elem = elem.nextElementSibling;
            }

            if (_circular && elem === null) {
                elem = getFirstElement(eTarget, _selector);
            }
        } else if (move === 'prev') {
            elem = eTarget.previousElementSibling;
            while (elem && !elem.matches(_selector)) {
                elem = elem.previousElementSibling;
            }

            if (_circular && elem === null) {
                elem = getLastElement(eTarget, _selector);
            }
        } else if (move === 'first') {
            elem = getFirstElement(eTarget, _selector);
        } else if (move === 'last') {
            elem = getLastElement(eTarget, _selector);
        }
    }

    if (elem && _focusElem) {
        elem.focus();
    }

    return elem;
}

/**
 * @param {NodeList} elemList List of elements to navigate through
 * @param {string} skipElemsSelector Skip elements that matching this selector
 * @param {number} startIdx Index into elemList
 * @param {'next'|'prev'} [find] Direction
 * @return {HTMLElement|null}
 */
export function findNextOrPrevElemByList({ elemList, skipElemsSelector, startIdx = -1, find = 'next' }) {
    const elemListLen = elemList.length;

    if (!skipElemsSelector || elemListLen === 0) {
        return null;
    }

    let el = elemList[startIdx] || null;
    let idx = startIdx;
    const findNext = find === 'next';

    while (el && el.matches(skipElemsSelector)) {
        if (findNext) {
            if (idx + 1 < elemListLen) {
                el = elemList[idx + 1];
                idx += 1;
            } else {
                el = null;
            }
        } else {
            if (idx - 1 >= 0) {
                el = elemList[idx - 1];
                idx -= 1;
            } else {
                el = null;
            }
        }
    }

    return el;
}

/**
 *
 * @param {KeyboardEvent} event
 * @param {NodeList} elemList List of elements to navigate through
 * @param {('horizontal' | 'vertical' | 'both')} [direction] Movement direction.
 * @param {string} [skipElemsSelector] Skip elements that matching this selector
 * @param {boolean} [circular] Circular keyboard navigation
 * @param {boolean} [focusElem] Focus found element.
 * @param {boolean} [useHomeAndEnd] `Home` and `End` keys are enabled.
 * @param {boolean} [addMissingTabIndex]
 * @return {HTMLElement|null} Next or previous element or null.
 */
export function keyboardNavigationByList({
    event,
    elemList,
    direction = 'horizontal',
    skipElemsSelector = '',
    circular = false,
    focusElem = true,
    useHomeAndEnd = true,
    addMissingTabIndex = true,
}) {
    if (!event || !(event.type in KEY_EVENTS)) {
        return null;
    }

    let elem = null;
    let move = '';
    let activeElem = null;
    let activeElemIdx = -1;
    const elemListLen = elemList.length;

    if (elemListLen > 0) {
        move = getMove({ event: event, direction: direction, useHomeAndEnd: useHomeAndEnd });

        if (move) {
            elemList.forEach((elem, idx) => {
                let tabindex = getAttr(elem, 'tabindex');

                if (addMissingTabIndex && !tabindex) {
                    setAttr(elem, 'tabindex', '-1');
                }

                if (tabindex === '0') {
                    activeElemIdx = idx;
                    activeElem = elem;
                }
            });

            if (activeElemIdx > -1) {
                if (move === 'next') {
                    if (activeElemIdx + 1 < elemListLen) {
                        elem = elemList[activeElemIdx + 1];
                    }

                    elem = findNextOrPrevElemByList({
                        elemList,
                        skipElemsSelector,
                        startIdx: activeElemIdx + 1,
                        find: 'next',
                    });

                    if (circular && elem === null) {
                        elem = findNextOrPrevElemByList({
                            elemList,
                            skipElemsSelector,
                            startIdx: 0,
                            find: 'next',
                        });
                    }
                } else if (move === 'prev') {
                    if (activeElemIdx - 1 >= 0) {
                        elem = elemList[activeElemIdx - 1];
                    }

                    elem = findNextOrPrevElemByList({
                        elemList,
                        skipElemsSelector,
                        startIdx: activeElemIdx - 1,
                        find: 'prev',
                    });

                    if (circular && elem === null) {
                        elem = findNextOrPrevElemByList({
                            elemList,
                            skipElemsSelector,
                            startIdx: elemListLen - 1,
                            find: 'prev',
                        });
                    }
                } else if (move === 'first') {
                    elem = findNextOrPrevElemByList({
                        elemList,
                        skipElemsSelector,
                        startIdx: 0,
                        find: 'next',
                    });
                } else if (move === 'last') {
                    elem = findNextOrPrevElemByList({
                        elemList,
                        skipElemsSelector,
                        startIdx: elemListLen - 1,
                        find: 'prev',
                    });
                }
            } else {
                activeElemIdx = 0;
                activeElem = elemList[0];
            }

            if (elem && activeElem) {
                setAttr(activeElem, 'tabindex', '-1');
                setAttr(elem, 'tabindex', '0');

                if (focusElem) {
                    elem.focus();
                }
            }
        }
    }

    return elem;
}
