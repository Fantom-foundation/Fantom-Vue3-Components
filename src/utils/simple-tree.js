import { getNestedProp } from './index.js';

export function isParent(_node) {
    return _node && _node._c && _node._c.length > 0;
}

export function walkTree(_nodes, _callback, _parent = null, _level = 0) {
    const callback = typeof _callback === 'function' ? _callback : null;
    let node;
    let action = '';

    if (_nodes && callback) {
        for (let i = 0, len1 = _nodes.length; i < len1; i++) {
            node = _nodes[i];
            action = callback(node, _parent, _level, i);

            if (action === 'stop') {
                return 'stop';
            } else if (action === 'skip') {
                continue;
            }

            if (isParent(node)) {
                if (walkTree(node._c, callback, node, _level + 1, i) === 'stop') {
                    return 'stop';
                }
            }
        }
    }

    return '';
}

export function findNodeBy(_nodes, _value, _prop = 'id') {
    let node = null;
    const parents = [];
    let level = 0;
    let index = 0;

    walkTree(_nodes, (_node, _parent, _level, _index) => {
        if (_level > level) {
            parents.push(_parent);
        } else if (_level < level) {
            let n = level - _level;

            while (n > 0) {
                parents.pop();
                n--;
            }

            level = _level;
        }

        level = _level;
        index = _index;

        if (getNestedProp(_node, _prop) === _value) {
            node = _node;
            return 'stop';
        }
    });

    return { node, parents, level, index };
}
