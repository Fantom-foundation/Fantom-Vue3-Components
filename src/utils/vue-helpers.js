import { getDefaultValue, getUniqueId } from './index.js';
import { nextTick } from 'vue';

export function getComponentName(_component) {
    return _component ? _component?.type?.name || _component?.$?.type?.name : '';
}

/**
 * @param {Object} obj
 * @param {string} action
 * @param {*} value
 * @return {Promise<unknown>}
 */
export function actionHandler(obj, action, value) {
    return new Promise((resolve) => {
        if (action in obj) {
            if (action === 'getState') {
                obj.states = [];
            }

            obj[action] = value;

            nextTick(() => {
                resolve();

                obj[action] = getDefaultValue(value);
            });
        }
    });
}

/**
 * Call function and wait for the next tick.
 *
 * @param {Function} fn
 * @return {Promise<unknown>}
 */
export function callAndWait(fn) {
    return new Promise((resolve) => {
        if (typeof fn === 'function') {
            fn();
        }

        nextTick(() => {
            resolve();
        });
    });
}

/**
 * @param {Object} component Vue component
 * @param {string} parentName
 * @return {Object|null} Vue component or null
 */
export function findParentByName(component, parentName) {
    let parent = component ? component.$parent : null;

    while (parent) {
        if (getComponentName(parent) === parentName) {
            return parent;
        }

        parent = parent.$parent;
    }

    return null;
}

export function valueOrNull(value) {
    return !value ? null : value;
}

/**
 * Generate unique id if there is no 'id' property in object.
 *
 * @param {Object[]} _items
 * @param {string} [_key]
 */
export function setIds(_items, _key = 'id') {
    let item;

    if (_items && _items.length) {
        for (let i = 0, len1 = _items.length; i < len1; i += 1) {
            item = _items[i];
            if (!(_key in item) || item[_key] === null) {
                item[_key] = getUniqueId();
            }
        }
    }
}

/**
 * Copy component's methods.
 *
 * @param {Object} component Vue component
 * @param {string[]} methods
 * @param {string} refName
 * @return {Object} Keys are method names, values are functions
 */
export function copyMethods(component, methods, refName = '') {
    const cMethods = {};
    const compMethods = component.methods;

    if (compMethods && methods) {
        methods.forEach((methodName) => {
            if (methodName in compMethods) {
                if (refName) {
                    cMethods[methodName] = function (...args) {
                        // this.$refs[refName][methodName](...args);
                        const ref = this.$refs[refName];

                        if (ref) {
                            compMethods[methodName].apply(ref, args);
                        }
                    };
                } else {
                    cMethods[methodName] = compMethods[methodName];
                }
            }
        });
    }

    return cMethods;
}

/**
 * Copy component's properties.
 *
 * @param {Object} component Vue component
 * @param {Object} props Keys are prop names, values are default values or null (default value won't be set)
 * @return {Object} Keys are prop names, values prop definition objects
 */
export function copyProps(component, props = {}) {
    const cProps = {};
    const compProps = component.props;

    if (compProps) {
        Object.keys(props).forEach((propName) => {
            const prop = props[propName];

            if (propName in compProps) {
                cProps[propName] = { ...compProps[propName] };

                if (prop !== null) {
                    // cProps[propName].default = clone(prop);
                    cProps[propName].default = prop;
                }
            }
        });
    }

    return cProps;
}

/**
 * Check for non-empty slot existence.
 *
 * @param {Object} component Vue component
 * @param {string} name
 * @return {boolean}
 */
export function hasSlot(component, name = 'default') {
    // return !!this.$slots[_name] || (this.$scopedSlots && !!this.$scopedSlots[_name]);
    return !!component.$slots[name];
}

/**
 * @param {Object} _component Vue component
 * @param {function} _callback
 * @return {boolean}
 */
/*
export function traverseVueTree(_component, _callback) {
    const children = _component.$slots ? _component.$slots.default() : null;
    let child;
    let goOn = true;

    if (children && children.length > 0) {
        for (let i = 0, len = children.length; i < len; i++) {
            child = children[i];

            if (_callback(child)) {
                goOn = traverseVueTree(child, _callback);
            } else {
                goOn = false;
                break;
            }
        }
    }

    return goOn;
}
*/

/**
 * Check the subtree of the vue component tree with the root `_startComponent`, if a component has `isChanged` method
 * and if it returns `true`.
 *
 * @param {Object} _startComponent Vue component
 * @param {Object} _messages Kyes are component names in kebab case, values are messages.
 * @param {boolean} [_noConfirmation] If `true`, no confirmation prompt will be raised
 * @returns {boolean|string} Wether to continue with an action or not.
 */
/*
export function isAnyComponentChanged(_startComponent, _messages = {}, _noConfirmation = false) {
    let changedComponentName = '';

    traverseVueTree(_startComponent, (_component) => {
        if ('isChanged' in _component && typeof _component.isChanged === 'function' && _component.isChanged()) {
            changedComponentName = getComponentName(_component);
            return false;
        }

        return true;
    });

    if (changedComponentName) {
        if (!_noConfirmation) {
            return !confirm(_messages[changedComponentName]);
        } else {
            return _messages[changedComponentName];
        }
    }

    return !!changedComponentName;
}
*/
