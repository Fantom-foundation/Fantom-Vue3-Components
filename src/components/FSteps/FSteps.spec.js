import { describe, it, expect, afterEach } from 'vitest';
import { mount } from '@vue/test-utils';
import FSteps from '@/components/FSteps/FSteps.vue';
import { destroyWrapper } from '@/test/utils.js';

let wrapper = null;

function createWrapper({ propsData = {}, slots = {} } = {}) {
    return mount(FSteps, {
        propsData,
        slots,
    });
}

afterEach(() => {
    destroyWrapper(wrapper);
});

describe('FSteps', () => {
    it('should render correctly', () => {
        wrapper = createWrapper();

        expect(wrapper.element).toMatchSnapshot();
    });

    it('should draw steps if `count` prop is given', async () => {
        wrapper = createWrapper({
            propsData: {
                count: 3,
            },
        });

        expect(wrapper.findAll('.fsteps_step').length).toBe(3);
    });

    it('should draw active steps if `active` prop is given', async () => {
        const activeClass = 'fsteps_step-active';

        wrapper = createWrapper({
            propsData: {
                count: 3,
                active: 2,
            },
        });

        const steps = wrapper.findAll('.fsteps_step');

        expect(steps[0].classes(activeClass)).toBeTruthy();
        expect(steps[1].classes(activeClass)).toBeTruthy();
        expect(steps[2].classes(activeClass)).toBeFalsy();
    });

    it('should draw labels if `labels` prop is given', async () => {
        wrapper = createWrapper({
            propsData: {
                count: 2,
                labels: ['step 1', 'step 2'],
            },
        });

        const labels = wrapper.findAll('.fsteps_label');

        expect(labels[0].text()).toBe('step 1');
        expect(labels[1].text()).toBe('step 2');
    });
});
