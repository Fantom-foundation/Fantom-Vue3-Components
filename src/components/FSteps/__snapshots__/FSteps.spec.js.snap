// Vitest Snapshot v1

exports[`FSteps > should render correctly 1`] = `
<div
  class="fsteps"
>
  <div
    class="fsteps_cont"
  >
    
    
  </div>
</div>
`;
